### PR TITLE
chore(sdk): Allow the launcher image to be configurable. 

### PR DESCRIPTION
**Pull request by neuromage**
**Description of your changes:**
Allow the launcher image to be configurable. Especially useful for tests.

**Checklist:**
- [X] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 
- [ ] Do you want this pull request (PR) cherry-picked into the current release branch?

    [Learn more about cherry-picking updates into the release branch](https://github.com/kubeflow/pipelines/blob/master/RELEASE.md#cherry-picking-pull-requests-to-release-branch).

### CODE REVIEW

File: sdk/python/kfp/compiler/compiler.py [hunk 1]
```diff
@@ -57,14 +57,23 @@ def my_pipeline(a: int = 1, b: str = "default value"):
 
   def __init__(
       self,
-      mode: dsl.PipelineExecutionMode = dsl.PipelineExecutionMode.V1_LEGACY):
+      mode: dsl.PipelineExecutionMode = dsl.PipelineExecutionMode.V1_LEGACY,
+      launcher_image: Optional[str] = None):
```
Comment (Ark-kun) [inline sdk/python/kfp/compiler/compiler.py:61]: Should this be the property the pipeline (`@dsl.pipeline(launcher_image=...)`)? Or maybe an attribute of the `kfp.dsl.PipelineConf` which is passed to `compiler.compile(pipeline_conf=...)`?

File: sdk/python/kfp/compiler/v2_compatible_compiler_test.py [hunk 2]
```diff
@@ -90,8 +88,28 @@ def v2_compatible_two_step_pipeline():
...
+          num_steps=preprocess_task.outputs['output_parameter_one'],
+          dataset=preprocess_task.outputs['output_dataset_one'])
+
+    kfp_compiler = compiler.Compiler(
+        mode=dsl.PipelineExecutionMode.V2_COMPATIBLE,
+        launcher_image='my-custom-image')
     self._assert_compiled_pipeline_equals_golden(
-        v2_compatible_two_step_pipeline, 'v2_compatible_two_step_pipeline.yaml')
+        kfp_compiler, v2_compatible_two_step_pipeline,
+        'v2_compatible_two_step_pipeline_with_custom_launcher.yaml')
```
Comment (Ark-kun) [inline sdk/python/kfp/compiler/v2_compatible_compiler_test.py:112]: JFYI: Such tests (based on golden compiled pipelines) are really fragile and add quite a lot of maintenance cost. They will start failing if any small thing is changed in the code generator or pipeline compiler.
We advice writing test cases that only check the required parts of the compiled pipeline. Namely the launcher image used in the templates.

File: sdk/python/kfp/compiler/compiler.py (hunk 1 above)
Comment (neuromage) [reply sdk/python/kfp/compiler/compiler.py:61]: No, I'd prefer this to be a property of the class itself for now.

File: sdk/python/kfp/compiler/v2_compat.py [hunk 3]
```diff
@@ -22,24 +23,30 @@
...
         runs.
       pipeline_root: The root output directory for pipeline artifacts.
+      launcher_image: An optional launcher image. Useful for tests.
     """
   # Inject the launcher binary and overwrite the entrypoint.
-  op.add_init_container(_LAUNCHER_CONTAINER)
+  image_name = launcher_image or _DEFAULT_LAUNCHER_IMAGE
+  launcher_container = dsl.UserContainer(name="kfp-launcher",
+                                         image=image_name,
+                                         command="/bin/mount_launcher.sh",
```
Comment (Ark-kun) [inline sdk/python/kfp/compiler/v2_compat.py:46]: Do we have any documentation regarding the launcher image interface 9e.g. it needs to have the `/bin/mount_launcher.sh` file with certain CLI interface).

File: sdk/python/kfp/compiler/v2_compatible_compiler_test.py (hunk 2 above)
Comment (neuromage) [reply sdk/python/kfp/compiler/v2_compatible_compiler_test.py:112]: Both kinds of tests seem reasonable to me.

File: sdk/python/kfp/compiler/v2_compat.py (hunk 3 above)
Comment (neuromage) [reply sdk/python/kfp/compiler/v2_compat.py:46]: Not yet. That will come when we've stabilized the feature.

File: sdk/python/kfp/compiler/compiler.py (hunk 1 above)
Comment (Ark-kun) [reply sdk/python/kfp/compiler/compiler.py:61]: I also wonder whether this should be per-task.
Does it make sense for different component tasks to use different launchers?

Comment (neuromage) [reply sdk/python/kfp/compiler/compiler.py:61]: Not right now.

**Review by Bobgy — COMMENTED** [review]
Thank you for the quick update!

Can we also add this option to https://github.com/kubeflow/pipelines/blob/d3ddfadf2dfbf99103bf02db5ec9511248399312/sdk/python/kfp/_client.py#L685-L693?

File: sdk/python/kfp/compiler/compiler.py [hunk 4]
```diff
@@ -57,14 +57,23 @@ def my_pipeline(a: int = 1, b: str = "default value"):
...
+    """Creates a KFP compiler for compiling pipeline functions for execution.
+
+    Args:
+      mode: The pipeline execution mode to use.
+      launcher_image: Configurable image for KFP launcher to use. Only applies
+        when `mode == dsl.PipelineExecutionMode.V2_COMPATIBLE`.
```
Comment (Bobgy) [inline sdk/python/kfp/compiler/compiler.py:67]: Shall we also mention that this field is only for dev/testing purposes?
(probably for users who want to deploy on-prem, they want to mirror official images, but other than that, users shouldn't set this image)

**Comment by Ark-kun** [comment]
> Can we also add this option to

Probably the best way to add it to the `PipelineConf` class which is intended to hold this king of customizations.

File: sdk/python/kfp/compiler/compiler.py (hunk 4 above)
Comment (neuromage) [reply sdk/python/kfp/compiler/compiler.py]: Done.

**Comment by neuromage** [comment]
> Thank you for the quick update!
> 
> Can we also add this option to
> 
> https://github.com/kubeflow/pipelines/blob/d3ddfadf2dfbf99103bf02db5ec9511248399312/sdk/python/kfp/_client.py#L685-L693
> 
> ?

Thanks, done.

**Comment by Bobgy** [comment]
Thank you!
/LGTM

**Pull request merged**